sycl/include/CL/sycl/detail/pi.hpp

```diff
@@ -117,68 +119,110 @@ void printArgs(Arg0 arg0, Args... args) {
+
+// Class to call PI API, trace, check the return result and throw Exception.
+// To Trace : Set SYCL_PI_TRACE environment variable.
+// Template Arguments:
+//    FnType, FnOffset - for CallPi Class.
+//    Exception - The type of exception to throw if PiResult of a call is not
+//    PI_SUCCESS. default is cl::sycl::runtime_error.
+// Usage:
+// Operator() - Call, Trace, check Result and Throw Exception.
+// Use Macro PI_CALL and PI_CALL_THROW to call the constructor directly.
+template <typename FnType, size_t FnOffset,
+          typename Exception = cl::sycl::runtime_error>
+class CallPiAndCheck : private CallPi<FnType, FnOffset> {
+public:
+  CallPiAndCheck() : CallPi<FnType, FnOffset>(){};
 
-} // namespace pi
+  template <typename... Args> void operator()(Args... args) {
+    PiResult Err = (CallPi<FnType, FnOffset>::operator()(args...));
+    checkPiResult<Exception>(Err);
+  }
+};
 
-namespace RT = cl::sycl::detail::pi;
+// Explicit specialization declarations for Trace class for every FnType.
+// The offsetof is used as a template argument to uniquely identify every
+// api.
+#define _PI_API(api)                                                           \
+  template <>                                                                  \
+  CallPi<decltype(&::api),                                                     \
+         (offsetof(pi_plugin::FunctionPointers, api))>::CallPi();
 
-#define PI_ASSERT(cond, msg) RT::assertion((cond), "assert: " msg);
+#include <CL/sycl/detail/pi.def>
 
-#define PI_TRACE(func) RT::Trace<decltype(func)>(func, #func)
+} // namespace pi
 
-// Use this macro to initialize the Plugin, call the API, do the trace 
-// and the check for no errors.
-#define PI_CALL(pi, ...)                                                       \
-  {                                                                            \
-    RT::initialize();                                                          \
-    RT::piCheckResult(PI_TRACE(pi)(__VA_ARGS__));                              \
-  }
+namespace RT = cl::sycl::detail::pi;
+
+// Use this macro to call the API, trace the call, check the return and throw a
+// runtime_error exception.
+// Usage: PI_CALL(pi)(Args);
+#define PI_CALL(pi)                                                            \
+  RT::CallPiAndCheck<decltype(&::pi),                                          \
+                     (offsetof(pi_plugin::FunctionPointers, pi))>()
+
+// Use this macro to call the API, trace the call and return the result.
+// To check the result use checkPiResult.
+// Usage:
+// PiResult Err = PI_CALL_NOCHECK(pi)(args);
+// RT::checkPiResult(Err); <- Checks Result and throws a runtime error
+// exception.
+#define PI_CALL_NOCHECK(pi)                                                    \
+  RT::CallPi<decltype(&::pi), (offsetof(pi_plugin::FunctionPointers, pi))>()
+
+// Use this macro to call the API, trace the call, check the return and throw a
```

Comment:
I think "throw a Exception" should be either "throw an Exception" or "throw the Exception".

Reply:
done.